packages/SwingSet/src/kernel/kernel.js

```diff
@@ -503,23 +510,93 @@ export default function buildKernel(kernelEndowments) {
     });
 
     // the vatManager invokes setup() to build the userspace image
-    const manager = makeVatManager(
+    return makeVatManager(
       vatID,
       syscallManager,
       setup,
       helpers,
       kernelKeeper,
-      vatKeeper,
-    );
-    ephemeral.vats.set(
-      vatID,
-      harden({
-        manager,
-        enablePipelining: Boolean(options.enablePipelining),
-      }),
+      kernelKeeper.provideVatKeeper(vatID),
     );
   }
 
+  // This kernel (and the genesis vats) will have been evaluated in a context in
+  // which these canonical symbols were available (which is why the require
+  // statements here work). That was achieved by running makeRequire() in
+  // controller.js. We want to provide the same context when we evaluate dynamic
+  // vats.
+  function kernelRequire(nameArg) {
+    const name = `${nameArg}`;
+    const HARDEN = '@agoric/harden';
+    const NAT = '@agoric/nat';
+    const EVALUATE = '@agoric/evaluate';
+    switch (name) {
+      case HARDEN:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(HARDEN);
+      case NAT:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(NAT);
+      case EVALUATE:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(EVALUATE);
+      default:
+        throw Error(`require ${name} is not supported`);
+    }
+  }
+
+  // Enqueue a message to the adminVat giving it the new vat's root object
+  function notifyAdminVatOfNewVat(vatID) {
+    function serializeSlot(slot) {
```

Comment:
maybe instead of defining `serializeSlot` and then calling it (once), we should just say `serializedArgs = { body: .. slots: [kernelRootObjectSlot] }`

I think in the other places where we use `serializeSlot`, it's a function that takes a single object (e.g. a Presence) and returns the `{@qclass:..}` bit that goes into the serialized body. Along the way, it modifies a larger-scale list of slots to include the appropriate object reference (`o+12`). Having `serializeSlot` used here to return the entire CapData object (which serializes an array of length one) is a bit non-parallel to the other uses.